style-guide.md

```diff
@@ -0,0 +1,65 @@
+# WDL Style Guide
+
+All rules below should be followed by contributors to this repo. Pull Requests which do not conform to these specifications will be asked to change.
+
+These rules might also be enforced by a yet-to-be-written linter.
+
+## Rules
+
+- The following sections must be present and in this order: `meta`, `parameter_meta`, `input`, `command`, `output`, `runtime`
+- The `meta` section should have a `description` of the task
+  - `meta` should *not* include `author` or `email` keys
+- All inputs must have a corresponding `parameter_meta` entry
+  - These texts should be copy and pasted from other tasks with the same input when possible
+  - Inputs and parameter meta must be in the same order
+  - TODO: rework this section
+  - That order should be logical
+    - look to similar tasks/workflows for an example of proper ordering
+    - Generally speaking:
+      - Required inputs at the top
+      - "sample" files before reference files
+      - Resource configuration at the bottom
+        - memory allocations come before disk space allocations
+      - `max_retries` should be last
+      - If applicable, `detect_nproc` immediately before `max_retries`
+- Tasks with string parameters for which a limited number of choices are valid, must be documented following the template in `string_choices_task` (see `template/task-templates.wdl`)
+  - TODO: this can't be enforced by a linter. Do we keep it in the guide?
+  - they should also fail quickly with an informative error message if an invalid input is provided
+    - In most cases, just passing the parameter to the underlying tool should produce a satisfactory error, but this must be checked for each task
+- All tasks must have configurable memory and disk space allocations
+  - TODO: only enforceable if we require specific var naming. Do we want to enforce specific variable names? Keep this in the guide?
+  - see the various tasks in the template directory for possible ways to allocate resources
+    - Contributors can mix and match the available templates, copy and pasting subsections as appropriate
+    - It is allowed to have one resource allocated dynamically, and another allocated statically in the same task.
+    - It is *not* allowed to have a resource which can be allocated *either* statically or dynamically.
+      - This is technically feasible, but is too complicated for maintenance and end-users.
+      - e.g. `memory_gb` and `modify_memory_gb` cannot be present in the same task.
```

Comment:
We could try to write some sort of pattern like `[0-9]+[GB|MB|KB]?`. That seems fraught with difficulty. The spec has:

```
Int - Intepreted as bytes
String - This should be a decimal value with suffixes like B, KB, MB or binary suffixes KiB, MiB. For example: 6.2 GB, 5MB, 2GiB. 
```

Reply:
I think this also belongs in the contributor guide. I think it would be too difficult to enforce. Because a check that "does an input fitting pattern `foo` exist?" isn't enough (and getting that pattern will be a challenge by itself). We would also need to check that it gets used in the allocation logic, which frequently involves multiple WDL statements. Technically possible to do, but IMO not worth implementing for v1 of a linter. We could come back to this in a potential v2 if we wanted.